### ISSUE TITLE
Help wanted about multiple edges

### ISSUE THREAD

**Issue by HongxuanZhang** [andrebrait/sinalgo#3]
Hello，it seems that according to the code in the file [sinalgo/nodes/NodeOutgoingConnectionsList.java](https://github.com/sinalgo/sinalgo/blob/master/src/main/java/sinalgo/nodes/NodeOutgoingConnectionsList.java),  there cannot be _multiple edges_ between two nodes, but I really need it in my project. Could you give me some advice about how to allow the  multiple edges without effect other functions of the framework? 

Thanks a lot!

**Comment by andrebrait** [comment]
Hi there,

May I ask you what kind of problem you're trying to simulate that requires multiple edges?

**Comment by HongxuanZhang** [comment]
Thanks for your reply. 

My network consists of many trees, and the network topology may change over time. 

For example, in two trees A and B, one node **α** may be the parent of another node **β** in the Tree A, but may be the child of **β** in the Tree B. At some time, the Tree B may have to change, and the node **β** could not be the parent of **α** in the Tree B after change. But the link between the node **α** and **β** have to remain since the **α** is still the parent of **β** in the Tree A.

So, it seems there should be two bidirectional links between the nodes **α** and **β**.

**Comment by andrebrait** [comment]
Hm, got it.
Let me think how this can be done and I'll get back to you.

Does one tree influence the other in ways that running the two simulations separately and then joining the results somehow is not enough? I'd assume so, but I'm asking just in case

**Comment by HongxuanZhang** [comment]
I really appreciate your help! 

Actually you don't have to think about the things about joining the results part, and in fact there is only one simulation but not two. I am only curious about how to use multiple edges based on your framework _sinalgo_.

Thanks for your any advice!

**Comment by HongxuanZhang** [comment]
Hi there! Have you got any idea?

**Comment by andrebrait** [comment]
Yes, I do.

So, what I think I can do is to allow you to have multiple, isolated sets of edges.
That way you could have each tree independently and, if situation calls for it, you can check for the connectivity between a given pair of nodes on each set of edges independently too.

Would that cover your use-case?

**Comment by HongxuanZhang** [comment]
Thanks,  that sounds really great!

May I ask how you implement it and how I can use it in my project?

**Comment by andrebrait** [comment]
Well, you'd have to modify Sinalgo's source code for that. 
The current plan is to make an array of lists of edges on the nodes and a few more changes, but it's more complicated than I can explain here. What are your deadlines on this?

**Comment by HongxuanZhang** [comment]
Well, the sooner the better. I have to finish this project in about one week. 

Many thanks for your work！Please tell me if I can help you to finish it faster.

**Comment by andrebrait** [comment]
Hm, a week is definitely not a timeframe in which I can implement this.

An idea would be to create two sets of nodes. Set A and set B, for example,
each containing the same number of nodes on the same coordinates. Then you
can have two trees at the same time. Some changes would have to be made to
the connectivity model, but I think it could work.

**Comment by HongxuanZhang** [comment]
Hello. I want to ask you whether this idea could make it.

In your code, the Node.java use NodeOutgoingConnectionsList which extends DoublyLinkedList as its Connections， but each entry is allowed only once in the collection according to the method _add(Node startNode, Node endNode, boolean valid)_. 

I think whether I can change it to a ConnectionsList which allow duplicate Edge, so when I need to modify one edge,  it would still have one copy in the List, which seems could meet my requirements.

**Comment by andrebrait** [comment]
I see no reason why you cannot do that, but you'd have to extend it
yourself.

One thing to be aware of is the `ConnectivityModel` you choose to use.
You'd have to handle node A being connected to B on a given tree, but not
on another, and check how you could add edges in each tree that way.
I think that, perhaps, simulating multiple trees by just creating a
duplicate of each node on top of each other node and treating them as
separate things when it comes to connectivity could be better, but I'm not
100% sure if that would work for you.

I'm very busy lately, and unfortunately I cannot look into the required
modifications myself.

André Brait

**Comment by HongxuanZhang** [comment]
Last question, should I put the class file which extends Node.java and allow duplicate Edges according to your advice in the folder named _nodeImplementations_ ？

Many thanks for your help!

**Comment by andrebrait** [comment]
Yes, that is correct.
The package also has to correspond to it, of course. Look into the examples
if you want to know the basic layout for it. Let me know if that's not
clear though.

**Comment by HongxuanZhang** [comment]
That 's clear, Thank you!
